### PR TITLE
[Android/iOS] Fix inconsistent behavior when setting StrokeDashArray

### LINKED ISSUE #12203: [Bug] Inconsistent behavior when setting StrokeDashArray values on an Ellipse
### Description
There is inconsistency between Android and iOS in how StrokeDashArray is interpreted and applied to an Ellipse shape.

### Steps to Reproduce
The following repro scenarios assume a single Ellipse shape has been added to a ContentPage. It has been entered horizontally and vertically, has a fixed height and width of 50, and has the Stroke and StrokeThickness Properties set to "Black" and "5" respectively.

Set the StrokeDashArray to:

1. Null
2. An empty DoubleCollection
3. A DoubleCollection with a single value greater than 0.0 e.g. 1.0
4. A DoubleCollection with two or more values greater than 0.0 e.g. { 1.0, 1.0 }
5. A DoubleCollection with two values, the first bring set to 0.0 and the second to 1.0

### Expected Behavior

Stroke appearance:

1. Solid (i.e. handle null/unsupported values as empty/default) or throw an exception
2. Solid 
3. Solid or Dashed (either but consistent across platforms)
4. Solid or Dashed (either but consistent across platforms)
5. Solid or Dashed (either but consistent across platforms)

The expectation is that the resulting appearance/behavior is consistent across platforms.

### Actual Behavior

Stroke appearance:

1. [INCONSISTENT] Android: throw exception, iOS: Solid
2. [CONSISTENT] Android: Solid, iOS: Solid 
3. [INCONSISTENT] Android: Solid, iOS: Dashed
4. [INCONSISTENT] Android: Solid, iOS: Dashed
5. [INCONSISTENT] Android: No Stroke, iOS: Dashed

### Basic Information

- Version with issue: 4.8.0.1364
- Last known good version:
- IDE: VS Mac
- Platform Target Frameworks: 
  - iOS:  13.7
  - Android: API 29  
- Android Support Library Version: 
- Nuget Packages:
  - Xamarin.Forms:  4.8.0.1364
  - Xamarin.Essentials:  1.5.3.2
- Affected Devices:
  - Honor 10 Lite:  API 29
  - iPhone X: 13.7

### Workaround

Use values that result in consistent behavior across all target platforms. Alternatively, use the OnPlatform markup extension to set the appropriate platform-specific value(s).

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change ###

Fix inconsistent behavior when setting **StrokeDashArray**.

### Issues Resolved ### 

- fixes #12203 

### API Changes ###
 
 None

### Platforms Affected ### 

- iOS
- Android

### Behavioral/Visual Changes ###

None

#### Before
Crash using a null value in Android.

#### After
Captura de pantalla 2020-09-25 a las 11 44 50

### Testing Procedure ###
Launch Core Gallery and navigate to the issue 11203. Verify all the Ellipses render in the same way on Android, iOS, etc.

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

**Review by rmarinho — APPROVED** [review]
Needs rebase, but it works

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Xamarin.Forms.Controls.Issues.Shared.projitems [hunk 1]
```diff
@@ -85,7 +85,7 @@
       <SubType>Code</SubType>
     </Compile>
     <Compile Include="$(MSBuildThisFileDirectory)Issue9006.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)Issue8207.xaml.cs" >
+    <Compile Include="$(MSBuildThisFileDirectory)Issue8207.xaml.cs">
```
Comment (hartez) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Xamarin.Forms.Controls.Issues.Shared.projitems:88]: Please revert all these whitespace changes so we can reduce conflicts.

File: Xamarin.Forms.Platform.Android/Extensions/DoubleCollectionExtensions.cs [hunk 2]
```diff
@@ -2,16 +2,28 @@
...
 
-            return array;
+                if (doubleCollection.Count % 2 == 0)
+                {
+                    array = new double[doubleCollection.Count];
+                    doubleCollection.CopyTo(array, 0);
+                }
+                else
+                {
+                    array = new double[2 * doubleCollection.Count];
```
Comment (hartez) [inline Xamarin.Forms.Platform.Android/Extensions/DoubleCollectionExtensions.cs:20]: Both of these methods need some sort of comment explaining why they create arrays which are twice as long if there is an odd number of elements.

File: Xamarin.Forms.Platform.Android/Shapes/ShapeRenderer.cs [hunk 3]
```diff
@@ -115,7 +115,15 @@ void UpdateStrokeThickness()
...
         void UpdateStrokeDashArray()
         {
-            Control.UpdateStrokeDashArray(Element.StrokeDashArray.ToArray());
+            var array = Element.StrokeDashArray.ToArray();
+
+            double thickness = Element.StrokeThickness;
+            var dashArray = new float[array.Length];
+
+            for (int i = 0; i < array.Length; i++)
+                dashArray[i] = (float)(thickness * array[i]);
```
Comment (hartez) [inline Xamarin.Forms.Platform.Android/Shapes/ShapeRenderer.cs:124]: Right now, the `ToArray` extension is creating an array which is then being copied again with some modifications (a cast to `float` and a scaling operation) - that means allocating an entire array which is immediately discarded. Is there any reason the cast and scaling can't be applied in the `ToArray` method, so we only allocate one extra array?

**Review by hartez — CHANGES_REQUESTED** [review]

File: Xamarin.Forms.Platform.iOS/Shapes/ShapeRenderer.cs [hunk 4]
```diff
@@ -111,34 +111,15 @@ void UpdateStrokeThickness()
...
-				}
-
-				double thickness = Element.StrokeThickness;
-
-                for (int i = 0; i < array.Length; i++)
-                    dashArray[i] = new nfloat(thickness * array[i]);
-                
-                Control.ShapeLayer.UpdateStrokeDash(dashArray);
-            }
+            var array = Element.StrokeDashArray.ToArray();
```
[the remark this answers is not in the document]
Comment (jsuarezruiz) [reply Xamarin.Forms.Platform.iOS/Shapes/ShapeRenderer.cs:114]: The `ToArray` extension method return an empty array.

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Xamarin.Forms.Controls.Issues.Shared.projitems (hunk 1 above)
Comment (jsuarezruiz) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Xamarin.Forms.Controls.Issues.Shared.projitems:88]: Reverted.

**Comment by jfversluis** [comment]
Needs changes which weren't implemented it seems, let's not do it for Forms anymore.

**Pull request closed without merging**